packages/loader/container-loader/src/container.ts

```diff
@@ -1364,20 +1417,30 @@ export class Container
 		const { snapshot, versionId } =
 			pendingLocalState === undefined
 				? await this.fetchSnapshotTree(specifiedVersion)
-				: { snapshot: undefined, versionId: undefined };
-		assert(
-			snapshot !== undefined || pendingLocalState !== undefined,
-			0x237 /* "Snapshot should exist" */,
+				: { snapshot: pendingLocalState.baseSnapshot, versionId: undefined };
+
+		if (pendingLocalState) {
+			this.baseSnapshot = pendingLocalState.baseSnapshot;
+			this.baseSnapshotBlobs = pendingLocalState.snapshotBlobs;
+		} else {
+			assert(snapshot !== undefined, 0x237 /* "Snapshot should exist" */);
+			if (this.offlineLoadEnabled) {
+				this.baseSnapshot = snapshot;
+				// Save contents of snapshot now, otherwise closeAndGetPendingLocalState() must be async
+				this.baseSnapshotBlobs = await getBlobContentsFromTree(snapshot, this.storage);
+			}
+		}
+
+		const attributes: IDocumentAttributes = await this.getDocumentAttributes(
+			this.storageAdapter,
+			snapshot,
 		);
 
-		const attributes: IDocumentAttributes =
-			pendingLocalState === undefined
-				? await this.getDocumentAttributes(this.storageAdapter, snapshot)
-				: {
-						sequenceNumber: pendingLocalState.protocol.sequenceNumber,
-						minimumSequenceNumber: pendingLocalState.protocol.minimumSequenceNumber,
-						term: pendingLocalState.term,
-				  };
+		// If we saved ops, we will replay them and don't need DeltaManager to fetch them
+		const sequenceNumber =
+			pendingLocalState?.savedOps[pendingLocalState.savedOps.length - 1]?.sequenceNumber;
+		const dmAttributes =
+			sequenceNumber !== undefined ? { ...attributes, sequenceNumber } : attributes;
```

Comment:
don't we need to fix up the minimum sequence number as well? as this could result in the sequenceNumber being below the minimum sequence number

Reply:
These should all be ops after the snapshot, so the sequence number should be >= the minimum sequence number

Reply:
do we even need this fix up then? should the snapshot always be at the right point? maybe an assert instead?

Reply:
i think the attributes should always be right, or we've missed saved ops, as snapshot is set to pendingLocalState.baseSnapshot if their is pending local state

Reply:
If we use the sequence number from the snapshot attributes, DeltaManager will download the ops we saved and we process them twice.

Reply:
ahh. ok. i get it now. thanks.